docs/core/diagnostics/metrics-collection.md

```diff
@@ -0,0 +1,404 @@
+    static Counter<int> s_hatsSold = s_meter.CreateCounter<int>(name: "hats-sold", 
+                                                                unit: "Hats",
+                                                                description: "The number of hats sold in our store"); 
+
+    static void Main(string[] args)
+    {
+        MeterListener meterListener = new MeterListener();
+        meterListener.InstrumentPublished = (instrument, listener) =>
+        {
+            if(instrument.Meter.Name == "HatCo.HatStore")
+            {
+                listener.EnableMeasurementEvents(instrument);
+            }
+        };
+        meterListener.SetMeasurementEventCallback<int>(OnMeasurementRecorded);
+        meterListener.Start();
+
+        Console.WriteLine("Press any key to exit");
+        while(!Console.KeyAvailable)
+        {
+            // Pretend our store has a transaction each second that sells 4 hats
+            Thread.Sleep(1000);
+            s_hatsSold.Add(4);
+        }
+
+        meterListener.Dispose();
+    }
+
+    static void OnMeasurementRecorded<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string,object>> tags, object? state)
+    {
+        Console.WriteLine($"{instrument.Name} recorded measurement {measurement}");
+    }
+}
+```
+
+When run the application now runs our custom callback on each measurement:
+```dotnetcli
+> dotnet run
+Press any key to exit
+hats-sold recorded measurement 4
+hats-sold recorded measurement 4
+hats-sold recorded measurement 4
+hats-sold recorded measurement 4
+...
+```
+
+Lets break down what happens in the example above. 
+```C#
+using var meterListener = new MeterListener();
+```
+First we created an instance of the <xref:System.Diagnostics.Metrics.MeterListener> which we will use to receive measurements.
+
+```C#
+meterListener.InstrumentPublished = (instrument, listener) =>
+{
+    if(instrument.Meter.Name == "HatCo.HatStore")
+    {
+        listener.EnableMeasurementEvents(instrument);
+    }
+};
+```
+Here we configured which instruments the listener will receive measurements from.
+<xref:System.Diagnostics.Metrics.MeterListener.InstrumentPublished> is a delegate that will be invoked any time a new
+instrument is created within the app. Our delegate can examine the instrument, such as checking the name, the Meter, or any other
+public property to decide whether to subscribe. If we do want to receive measurements from this instrument then we invoke
+<xref:System.Diagnostics.Metrics.MeterListener.EnableMeasurementEvents> to indicate that. If your code has another way
+to obtain a reference to an instrument it is legal to invoke EnableMeasurementEvents() at any time with that reference, but this is
+probably uncommon.
+
+```C#
+meterListener.SetMeasurementEventCallback<int>(OnMeasurementRecorded);
+...
+static void OnMeasurementRecorded<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string,object>> tags, object? state)
```

Comment:
>T [](http://example.com/codeflow?start=34&length=1)

If you decide to change this to `int` in the sample, you'll need to change it here too.

Reply:
I wanted to leave it as T because so that it easier for users to extend the sample with multiple calls to SetMeasurementEventCallback() on different data types. I expect only needing a single data type will be uncommon.

Reply:
This is reply to my other comment :-) don't you think it would be better for the perf to use non generic type? I am ok to leave it as `T` but add a small comment the user can provide non generic type too which would be perform better?

Reply:
> don't you think it would be better for the perf to use non generic type?

In this case, no I don't expect there is a perf difference but I'll admit I didn't do benchmarking to confirm that I didn't overlook something. I'm not planning to go benchmark it now, but if someone else did that and it turns out I had it wrong I'm happy to change it.